pkg/storage/azure/azureclient/azureclient.go

```diff
@@ -102,47 +102,51 @@ func (c *Client) getCreds() (azcore.TokenCredential, error) {
-		options := azidentity.ManagedIdentityCredentialOptions{
-			ClientOptions: azcore.ClientOptions{
-				Cloud: c.clientOpts.Cloud,
-			},
-			ID: azidentity.ClientID(managedIdentityClientID),
-		}
-		creds, err = azidentity.NewManagedIdentityCredential(&options)
-		if err != nil {
-			return nil, err
+	// For each of the following configuration values, if the environment variable is already set, use it over any
+	// configuration value: AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET. Otherwise, set the environment
+	// variable to the configuration value. The Azure SDK for Go function, NewDefaultAzureCredential, checks environment
+	// variables first.
+	if os.Getenv(util.AzureTenantID) == "" {
+		if err = os.Setenv(util.AzureTenantID, strings.TrimSpace(c.opts.TenantID)); err != nil {
+			return nil, fmt.Errorf("failed to set environment variable %s: %w", util.AzureTenantID, err)
 		}
-	} else if strings.TrimSpace(c.opts.ClientSecret) == "" {
-		options := azidentity.WorkloadIdentityCredentialOptions{
-			ClientOptions: *c.clientOpts,
-			ClientID:      c.opts.ClientID,
-			TenantID:      c.opts.TenantID,
-			TokenFilePath: c.opts.FederatedTokenFile,
+	}
+
+	if os.Getenv(util.AzureClientID) == "" {
+		if err = os.Setenv(util.AzureClientID, strings.TrimSpace(c.opts.ClientID)); err != nil {
+			return nil, fmt.Errorf("failed to set environment variable %s: %w", util.AzureClientID, err)
 		}
-		creds, err = azidentity.NewWorkloadIdentityCredential(&options)
-		if err != nil {
-			return nil, err
+	}
+
+	if os.Getenv(util.AzureClientSecret) == "" {
+		if err = os.Setenv(util.AzureClientSecret, strings.TrimSpace(c.opts.ClientSecret)); err != nil {
+			return nil, fmt.Errorf("failed to set environment variable %s: %w", util.AzureClientSecret, err)
 		}
-	} else {
-		options := azidentity.ClientSecretCredentialOptions{
-			ClientOptions: *c.clientOpts,
+	}
+
+	if os.Getenv(util.AzureFederatedTokenFile) == "" {
+		if err := os.Setenv(util.AzureFederatedTokenFile, strings.TrimSpace(c.opts.FederatedTokenFile)); err != nil {
+			return nil, fmt.Errorf("failed to set environment variable %s: %w", util.AzureFederatedTokenFile, err)
 		}
-		creds, err = azidentity.NewClientSecretCredential(
-			c.opts.TenantID,
-			c.opts.ClientID,
-			c.opts.ClientSecret,
-			&options,
-		)
-		if err != nil {
-			return nil, err
+	}
+
+	if os.Getenv(util.AzureClientSecret) == "" && os.Getenv(util.AzureFederatedTokenFile) == "" {
+		if err = os.Setenv(util.AzureFederatedTokenFile, "/var/run/secrets/openshift/serviceaccount/token"); err != nil {
+			return nil, fmt.Errorf("failed to set environment variable %s: %w", util.AzureFederatedTokenFile, err)
 		}
 	}
-	if creds == nil {
-		return nil, errors.New("Unknown authentication method")
+
+	options := &azidentity.DefaultAzureCredentialOptions{
+		ClientOptions: azcore.ClientOptions{
+			Cloud: c.clientOpts.Cloud,
+		},
+	}
+
+	creds, err = util.AzureCreds(options)
+	if err != nil {
+		return nil, err
 	}
+
```

Comment:
I'm hesitant to merge this - it's changes every flavor of auth in the azure driver, which is very invasive and bound to break _something_.

Relatedly, it doesn't look like `NewDefaultAzureCredential` supports `ClientSecretCredential` ? Source: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/default_azure_credential.go#L39-L57

Reply:
It does. It's just buried down in the environment credentials:
- https://github.com/Azure/azure-sdk-for-go/blob/8759ee622df98a0f205a05de9fc558d3ad0c7d65/sdk/azidentity/default_azure_credential.go#L77
- https://github.com/Azure/azure-sdk-for-go/blob/8759ee622df98a0f205a05de9fc558d3ad0c7d65/sdk/azidentity/environment_credential.go#L102